concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,184 @@
+        if *heap_end == 0 {
+            let actual_size = self.size().size_in_bytes();
+            *heap_end = actual_size;
+        }
+
+        // Align the address.
+        let alloc_start = align_up(*next, layout.align());
+        // Get the end of the allocation. This should always return `Some` as the
+        // contract memory is limited to much below usize::MAX.
+        let alloc_end = match alloc_start.checked_add(layout.size()) {
+            Some(end) => end,
+            None => return ptr::null_mut(),
+        };
+
+        // Check if we need to request more memory.
+        if alloc_end > *heap_end {
+            let space_needed = alloc_end - *heap_end;
+            let pages_to_request = PageCount(div_ceil(space_needed, PAGE_SIZE));
+            let previous_page_count = self.grow_memory(pages_to_request);
+            // Check if we are out of memory.
+            if previous_page_count == ERROR_PAGE_COUNT {
+                return ptr::null_mut();
+            }
+            // Increase the heap size.
+            *heap_end += pages_to_request.size_in_bytes();
+        }
+
+        // Increment allocations counter.
+        *self.allocations.get() += 1;
+        // Remember the last address handed out, so that we may move next backwards if
+        // it is deallocated before a new allocation occurs.
+        *self.last_alloc.get() = alloc_start;
+        *next = alloc_end;
+        alloc_start as *mut u8
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        let allocations = self.allocations.get();
+        let last_alloc = self.last_alloc.get();
+        let next = self.next.get();
+        // Decrease the allocation counter.
+        *allocations -= 1;
+        // Reset next and last allocation if everything has been deallocated.
+        if *allocations == 0 {
+            *next = MIN_PTR_ADDR;
+            *last_alloc = MIN_PTR_ADDR;
+        } else if *last_alloc as *mut u8 == ptr {
+            // Move next back to last alloc. This is a small optimization over the regular
+            // bump allocator.
+            *next = *last_alloc;
+        }
+    }
+}
+
+/// Align the given address, `addr`, upwards to alignment `align`.
+///
+/// Requires that `align` is a power of two, which it always is due to the
+/// design of [`Layout`].
+///
+/// Uses a bitmask to align the addresses for efficiency.
+/// For details, see: https://os.phil-opp.com/allocator-designs/#address-alignment
+fn align_up(addr: usize, align: usize) -> usize { (addr + align - 1) & !(align - 1) }
+
+/// Calculates the quotient of `lhs / rhs`, while rounding up to the nearest
+/// integer.
+///
+/// ```
+/// let a = 8;
+/// let b = 3;
+/// assert_eq!(a.div_ceil(b), 3);
+/// assert_eq!(a.div_ceil(-b), -2);
+/// assert_eq!((-a).div_ceil(b), -2);
+/// assert_eq!((-a).div_ceil(-b), 3);
+/// ```
+const fn div_ceil(lhs: usize, rhs: usize) -> usize {
+    let d = lhs / rhs;
+    let r = lhs % rhs;
```

Comment:
Final comment, since this is only used when rhs = PAGE_SIZE, I think in that special case this would be
```
lhs >> 16 + (lhs & 0xffff != 0 as usize)
```

Reply:
I tried out your proposed solution Ales, but it seemed off by one in a few cases. It also isn't the easiest thing to read, and the compiler may end up doing something similar on its own.
But as you pointed out, the `rhs` is constant, so I created a more bespoke function for calculating it.

Feel free to add the bitwise arithmetic if you find it essential.

Reply:
Hmm, did you use what I intended rather than what I wrote (the parenthesis)
```
fn f1(xs: usize) -> usize {
    (xs >> 16) + ((xs & 0xffff != 0) as usize)
}
```
This program
```
fn f1(xs: usize) -> usize {
    (xs >> 16) + ((xs & 0xffff != 0) as usize)
}

const N: usize = 1 << 16;

fn f2(xs: usize) -> usize {
    let a = xs / N;
    if xs % N > 0 {
        a + 1
    } else {
        a
    }
}

fn main() {
    for x in 0..usize::MAX {
        assert_eq!(f1(x), f2(x), "Failed for {x}");
    }
}
```
Does not panic.

Reply:
Ah, turns out I had `PAGE_SIZE` set to `65535` in the Rust playground instead of `65536`, which caused the discrepancy. I've used your suggested approach now ;)

Reply:
:)